pkg/edvprovider/edvprovider.go

```diff
@@ -105,30 +135,29 @@ func (c *Provider) openUnderlyingStore(vaultID string) (underlyingStore storage.
 			fmt.Errorf("failed to determine underlying store name: %w", err)
 	}
 
-	coreStore, err := c.coreProvider.OpenStore(storeName)
+	coreStore, err := c.CoreProvider.OpenStore(storeName)
 	if err != nil {
 		return nil, "", err
 	}
 
 	return coreStore, storeName, nil
 }
 
-// Store represents an EDV store.
-// It wraps an Aries store with additional functionality that's needed for EDV operations.
-type Store struct {
-	coreStore         storage.Store
-	name              string
-	retrievalPageSize uint
-}
-
-// Put stores the given document.
-func (c *Store) Put(document models.EncryptedDocument) error {
-	return c.UpsertBulk([]models.EncryptedDocument{document})
+// Vault represents a single vault store.
+type Vault struct {
+	CoreStore           storage.Store
+	underlyingStoreName string
+	retrievalPageSize   uint
 }
 
-// UpsertBulk stores the given documents, creating or updating them as needed.
+// Put stores the given documents, creating or updating them as needed.
 // TODO (#236): Support "unique" option on attribute pair.
-func (c *Store) UpsertBulk(documents []models.EncryptedDocument) error {
+func (v *Vault) Put(documents ...models.EncryptedDocument) error {
+	mongoDBStore, ok := v.CoreStore.(*mongodb.Store)
```

Comment:
It might be cleaner to create two different implementations of this function, i.e. one for MongoDB and one for the rest. For example:

`
type Vault struct {
  Put func(documents ...models.EncryptedDocument) error
}
`

When you open the vault you can check if it's a MongoDB store and then set the impl, e.g.

`
  vault.Put = vault.mongoDBPut
`

Just a suggestion :)